exercises/code-scanning.md

```diff
@@ -318,5 +318,89 @@ queries:
 ```
 </details>
 
+#### _Stretch Exercise 4a: Uploading the SARIF as a workflow artifact_
+    
+The output of the `github/codeql-action/analyze@v1` is a SARIF. What action should we use to upload the SARIF as an artifact?
+<details>
+<summary>Solution</summary>
+
+```yaml 
+    - name: Perform CodeQL Analysis
+      uses: github/codeql-action/analyze@v1
+      with:
+        output: code-scanning-results
+
+    - name: Upload SARIF as a Build Artifact
+      uses: actions/upload-artifact@v2
+      with:
+        name: sarif
+        path: code-scanning-results
+        retention-days: 7
+```
+</details>
+    
+#### _Stretch Exercise 4b: Uploading CodeQL databases as workflow artifacts_
+    
+By looking at the logs, where does CodeQL output the CodeQL databases, and similar to the previous exercise, how do we upload this?
+    
+
+**Hints**
+- [How to set outputs of a step](https://github.com/actions/toolkit/blob/main/docs/commands.md)
+- [CodeQL version in ubuntu-latest GitHub hosted runner](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#tools)
+- [CodeQL CLI Reference](https://codeql.github.com/docs/codeql-cli/manual/)
+  - [codeql database bundle](https://codeql.github.com/docs/codeql-cli/manual/database-bundle/)
+
+<details>
+<summary>Solutions</summary>
+
+```yaml 
+    - name: Upload CodeQL database
+      id: codeql-database-bundle
+      env:
+        LANGUAGE: ${{ matrix.language }}
+        CODEQL_PATH: /opt/hostedtoolcache/CodeQL/<codeql-bundle-name>/x64/codeql/codeql
```

Comment:
I would maybe mention how to explicitly find the `<codeql-bundle-name>` part of the path
and/or highlight that is an option that needs to be filled in for the solution.

Unless its actually super obvi from the [doc](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#tools) you listed above, but the way I found its value was in the sarif artifact I could see `/opt/hostedtoolcache/CodeQL/0.0.0-20210622/x64/codeql/qlpacks/codeql-go-examples/` which means the bundle name being used is `0.0.0-20210622` in my runner, and filling in that value enables this step to work, but finding that was... maybe not totally intuitive. Also possible I took the convoluted route though!